feat: rpc to delete verifiable credentials (#492)

AndrewCLu opened the pull request:
## Explanation

This change adds two RPC endpoints: DELETE_VERIFIABLE_CREDENTIAL and DELETE_ALL_VERIFIABLE_CREDENTIALS. The former deletes a Verifiable Credential from the user's storage via a specific `credential_id`. The latter deletes all Verifiable Credentials for a given user.

Note: This change is stacked on top of [data types](https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/494) and [credentials service](https://github.com/CryptKeeperZK/crypt-keeper-extension/pull/535) PRs.
- [x] RPC endpoint to delete a Verifiable Credential by Id
- [x] RPC endpoint to delete all Verifiable Credentials
- [x] Integration with History and Notifications services
- [x] Register RPC endpoints
- [x] Tests for deleting Verifiable Credentials

## Pre-Merge Checklist

- [x] PR template is filled out
- [ ] **IF** this PR fixes a bug, a test that _would have_ caught the bug has been added
- [x] PR is linked to the appropriate GitHub issue
- [x] PR has been added to the appropriate release Milestone

> PR template source from [github.com/MetaMask](https://github.com/MetaMask)

0xmad commented on packages/app/src/util/credential.ts:
```diff
@@ -0,0 +1,205 @@
+  }
+
+  if (
+    obj.issuanceDate &&
```
we have `yup` schema validation so you can write rules without adding custom validation mechanism.
This is related to all functions in this file.

AndrewCLu replied:
updated to use yup schema validation

0xmad commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -0,0 +1,176 @@
+    verifiableCredential: VerifiableCredential,
+  ): Promise<boolean> => {
+    /**
+     * We do not accept Verifiable Credentials that do not have a non-empty id.
```
If we don't use empty fields maybe we can update the types and make them stricter as we discussed before?

AndrewCLu replied:
Upon doing more research, it seems like some credential schemes (specifically [anonymous credentials](https://hyperledger.github.io/anoncreds-spec/#introduction)) avoid using the id parameter in a credential for linkability purposes. So I think it makes sense for us to use the minimal specification for a Verifiable Credential when storing. That being said, I will have to modify how we index into a stored credential, perhaps by using a hash.

I think enforcing strict types may create some compatibility issues that will bug us in the future.

0xmad commented on packages/app/src/background/services/credentials/__tests__/credentials.test.ts:
```diff
@@ -0,0 +1,201 @@
+    },
+  };
+  const exampleCredentialString = JSON.stringify(exampleCredential);
+  const exampleCredential2: VerifiableCredential = {
```
Let's get rid of number in var names

AndrewCLu replied:
fixed

0xmad reviewed (changes requested):
@AndrewCLu thanks just a few moments need to be resolved.

0xmad commented on packages/app/src/util/credential.ts:
```diff
@@ -0,0 +1,136 @@
+      .mixed()
+      .required()
+      .test("is-string-or-CredentialIssuer", "issuer must be a string or CredentialIssuer", (value) => {
+        if (typeof value === "string") {
+          return true;
+        }
+        try {
+          CredentialIssuerSchema.validateSync(value);
+          return true;
+        } catch (error) {
+          return false;
+        }
```
```suggestion
        if (typeof value === "string") {
          return true;
        }

        try {
          CredentialIssuerSchema.validateSync(value);
          return true;
        } catch (error) {
          return false;
        }
```

0xmad commented on packages/app/src/util/credential.ts:
```diff
@@ -0,0 +1,136 @@
+    expirationDate: yup.date().transform(parseDate).optional(),
+    credentialSubject: CredentialSubjectSchema,
+    credentialStatus: yup.lazy((value) => {
+      if (typeof value === "undefined") {
+        return yup.object().optional();
+      }
+      return CredentialStatusSchema.required() as yup.Schema;
```
```suggestion
      if (!value) {
        return yup.object().optional();
      }
 
      return CredentialStatusSchema.required() as yup.Schema;
```

AndrewCLu commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -0,0 +1,189 @@
+    await this.notificationService.create({
+      options: {
+        title: "Verifiable Credential deleted.",
+        message: `Credential Id: ${credentialId}`,
```
I removed the id from the notification. There's not really a notion of credential name to display, unless say, the user would like to nickname their credentials within the extension

0xmad replied:
Yes, we have the same logic for identities and this data is located in metadata.

AndrewCLu replied:
Sounds good. Rewrote logic for credentials using metadata, see below

0xisk commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -0,0 +1,205 @@
+    return VerifiableCredentialsService.INSTANCE;
+  }
+
+  addVerifiableCredential = async (verifiableCredentialJson: string): Promise<boolean> => {
```
```suggestion
  addVerifiableCredential = async (serializedVerifiableCredential: string): Promise<boolean> => {
```

0xisk commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -0,0 +1,205 @@
+    );
+  };
+
+  deleteVerifiableCredential = async (credentialId: string): Promise<boolean> => {
```
```suggestion
  deleteVerifiableCredential = async (verifiableCredentialId: string): Promise<boolean> => {
```

0xisk commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -0,0 +1,205 @@
+      return false;
+    }
+
+    const credentials = await this.getVerifiableCredentialsFromStore();
```
```suggestion
    const verifiableCredentials = await this.getVerifiableCredentialsFromStore();
```

0xisk commented on packages/app/src/background/services/credentials/index.ts:
```diff
@@ -0,0 +1,205 @@
+
+    const verifiableCredential = await parseVerifiableCredentialFromJson(verifiableCredentialJson);
+
+    if (verifiableCredential === null) {
```
This condition wont be needed it we throw errors in `validateVerifiableCredential()` function.

0xisk reviewed (changes requested):
Thanks @AndrewCLu, I left some comments.

AndrewCLu commented on packages/app/src/util/credential.ts:
So I originally thought there might be some use for VC utils outside of the background, but I can't think of anything immediately, so I moved it to the service as you suggested

AndrewCLu commented:
@0xisk @0xmad Doing some more thinking, I think it makes sense to store Verifiable Credentials with metadata, just as we do for Identities. 

- I've updated my changes to include a `CryptkeeperVerifiableCredential` type, which encompasses the metadata (open to naming suggestions). Metadata right now includes the name that the user assigns to the credential, as well as its hash.
- Going along with this, I also updated the storage of VCs to use hashes, so we are no longer dependent on credentials having ids (they are an optional property in the standard, and are avoided for anonymous credentials). 
- `VerifiableCredential` is moved to a workspace-level type, as for example they will be used for the demo. `CryptkeeperVerifiableCredential` is a package specific type.
- I renamed/refactored all the scattered JSON parsing and stringifying into a set of serialization/deserialization functions in the utils file, which I think makes the code easier to understand.

Please let me know your feedback, thanks.

0xmad commented on packages/app/src/background/services/credentials/utils.ts:
```diff
@@ -0,0 +1,243 @@
+  return validateVerifiableCredential(deserializedVerifiableCredential);
+}
+
+let claimValueSchema: yup.Schema<ClaimValue>;
```
You can get rid of let and use const with variable initialization.

AndrewCLu replied:
The use of let here addresses a circular dependency naming issue that comes when `claimValueSchema` is not declared before the body of `claimValueMapSchema`

0xmad replied:
Ok, this one works for now. In the future we can use class to get rid of hoisting.

0xmad commented on packages/app/src/background/services/credentials/utils.ts:
```diff
@@ -0,0 +1,243 @@
+ */
+export async function validateVerifiableCredential(verifiableCredential: object): Promise<VerifiableCredential> {
+  try {
+    return await verifiableCredentialSchema.validate(verifiableCredential);
```
```suggestion
    return verifiableCredentialSchema.validate(verifiableCredential);
```

AndrewCLu replied:
This await waits for the promise to resolve so that the error will actually be caught by the try-catch

0xmad replied:
Ok, in this case you can remove try/catch. The error will be rejected anyway.

AndrewCLu replied:
Ah oops, good catch

0xmad commented on packages/app/package.json:
```diff
@@ -116,6 +117,7 @@
     "eventemitter2": "^6.4.9",
     "extension-port-stream": "^2.1.1",
     "fast-deep-equal": "^3.1.3",
+    "json-stable-stringify": "^1.0.2",
```
Actually, I don't trust such packages and relying on string representation of object can be danger due to compatibility and collisions. I think we can use https://docs.ethers.org/v5/api/utils/hashing/#TypedDataEncoder but it's still in experimental stage so if there is another option, let's discuss.

AndrewCLu replied:
Ok, yes I couldn't find any other options, as you mentioned the one you linked is experimental. Will explore other options but the package I found seemed to be the most widely used/stable

AndrewCLu commented:
> @AndrewCLu thanks, just left some comments and have one question. If user updates metadata for credential, will it update the hash as well?

No, the hash is just of the credential itself, almost like an identifier. This is why we need consistent serialization for credentials, to produce consistent identifiers.

AndrewCLu commented on packages/app/src/background/services/credentials/utils.ts:
```diff
@@ -0,0 +1,243 @@
+    throw new Error("Serialized Cryptkeeper Verifiable Credential is not provided!");
+  }
+
+  let parsedCryptkeeperVerifiableCredential;
+  try {
+    parsedCryptkeeperVerifiableCredential = JSON.parse(serializedCryptkeeperVerifiableCredential) as {
+      verifiableCredential: string;
+      metadata: VerifiableCredentialMetadata;
+    };
+  } catch (error) {
+    throw new Error(`Serialized Cryptkeeper Verifiable Credential is not valid JSON: ${(error as Error).message}`);
+  }
+
+  return {
+    verifiableCredential: await deserializeVerifiableCredential(
+      parsedCryptkeeperVerifiableCredential.verifiableCredential,
+    ),
+    metadata: parsedCryptkeeperVerifiableCredential.metadata,
+  };
```
The use of the `let` statement is to separate the errors between `failure to parse JSON` and `failure to deserialize Verifiable Credential`. I don't want to catch a deserialization error and throw a JSON parsing error, as I feel like these are separate errors.

0xmad replied:
Ok, in this case you can just get rid of try/catch and get default parsing error.

AndrewCLu replied:
Makes sense, fixed

AndrewCLu commented on packages/app/src/background/services/credentials/utils.ts:
```diff
@@ -0,0 +1,243 @@
+    throw new Error("Serialized Verifiable Credential is not provided!");
+  }
+
+  let deserializedVerifiableCredential;
+  try {
+    deserializedVerifiableCredential = JSON.parse(serializedVerifiableCredential) as Record<string, unknown>;
+  } catch (error) {
+    throw new Error(`Serialized Verifiable Credential is not valid JSON: ${(error as Error).message}`);
+  }
+
+  return validateVerifiableCredential(deserializedVerifiableCredential);
```
Same as above, I wanted to separate JSON parsing from deserialization errors

0xmad reviewed (approved):
@AndrewCLu thanks!

0xisk commented:
@AndrewCLu  thanks, lets make sure to squash and merge.

0xisk commented on packages/app/src/background/services/credentials/__tests__/utils.test.ts:
```diff
@@ -0,0 +1,630 @@
+});
+
+describe("util/validateVerifiableCredential", () => {
+  test("should correctly validiate a valid verifiable credential", async () => {
```
```suggestion
  test("should correctly validate a valid verifiable credential", async () => {
```
